### PR TITLE
minor Span impl improvements, no locks on atomic assignments

### LINKED ISSUE #271: Lock in Span SDK  should be revisited
Concurrency on Span APIs is probably not a real-world case.

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #271 

**[update] No perf improvements are coming because Span has to be thread-safe, ignore numbers below.**

In the majority of cases, Span is not used concurrently. Even when attributes or events are added, it does not normally happen concurrently.

Let's declare span as not thread-safe and let rare instrumentations that need concurrency do the locking.
If we'll find it's common with e.g. Events, we can create lazy lock when first event is added.

this change removes locking which this simplifies code and reduces allocations 

**After**

|                           Method |        Mean |      Error |     StdDev | Allocated |
|--------------------------------- |------------:|-----------:|-----------:|----------:|
|               CreateSpan_Sampled |   772.27 ns |  7.1575 ns |  6.6951 ns |     560 B |
|    CreateSpan_Attributes_Sampled |   929.65 ns |  8.1989 ns |  7.6692 ns |    1248 B |
|     CreateSpan_Propagate_Sampled | 1,859.91 ns | 19.3340 ns | 17.1391 ns |     933 B |
| CreateSpan_Attributes_NotSampled |   795.95 ns |  7.8778 ns |  7.3689 ns |     560 B |
|                  CreateSpan_Noop |    38.63 ns |  0.3342 ns |  0.2962 ns |         - |
|       CreateSpan_Attributes_Noop |    44.45 ns |  0.3592 ns |  0.3360 ns |         - |
|        CreateSpan_Propagate_Noop |    41.12 ns |  0.4033 ns |  0.3773 ns |         - |

**Before**

|                           Method |        Mean |      Error |     StdDev | Allocated |
|--------------------------------- |------------:|-----------:|-----------:|----------:|
|               CreateSpan_Sampled |   775.89 ns | 10.9065 ns |  9.6683 ns |     592 B |
|    CreateSpan_Attributes_Sampled | 1,016.07 ns |  6.3040 ns |  5.5883 ns |    1280 B |
|     CreateSpan_Propagate_Sampled | 1,894.26 ns | 22.0567 ns | 20.6319 ns |     966 B |
| CreateSpan_Attributes_NotSampled |   802.46 ns |  5.3171 ns |  4.9736 ns |     592 B |
|                  CreateSpan_Noop |    38.56 ns |  0.2433 ns |  0.2275 ns |         - |
|       CreateSpan_Attributes_Noop |    44.65 ns |  0.3581 ns |  0.3350 ns |         - |
|        CreateSpan_Propagate_Noop |    41.41 ns |  0.3090 ns |  0.2891 ns |         - |

### CODE REVIEW

File: src/OpenTelemetry/Trace/Span.cs [hunk 1]
```diff
@@ -174,92 +159,79 @@ public Status Status
...
+            }
+
             this.Name = name ?? throw new ArgumentNullException(nameof(name));
         }
 
         /// <inheritdoc/>
         public void SetAttribute(KeyValuePair<string, object> keyValuePair)
         {
-            if (keyValuePair.Key == null || keyValuePair.Value == null)
+            if (keyValuePair.Key == null)
```
Comment (pcwiese) [inline src/OpenTelemetry/Trace/Span.cs]: A general comment about all of the parameter validations we are putting in.

Seems to violate the basic error handling principles outlined here:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/error-handling.md

No?

**Comment by pcwiese** [comment]
Do you need to explicitly say that Span is not thread-safe in the class documentation? Should that documentation be on ISpan itself in the API?

File: src/OpenTelemetry/Trace/Span.cs (hunk 1 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Span.cs]: Cool, I didn't know that we have error handling principles. You are right.  This is much wider that Span only, I created a new issue to track it https://github.com/open-telemetry/opentelemetry-dotnet/issues/295.

**Comment by lmolkova** [comment]
It seems I sent this PR without reading the spec: 

["All methods of Span are safe to be called concurrently"}(https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/concurrency.md)

So Span must be thread safe 

I'll update this PR to keep some of its goodness and return lock back

File: src/OpenTelemetry/Trace/Span.cs (hunk 1 above)
Comment (SergeyKanzhelev) [reply src/OpenTelemetry/Trace/Span.cs]: thanks!

**Pull request merged**